packages/neon_http_client/lib/src/interceptors/http_interceptor.dart

```diff
@@ -0,0 +1,26 @@
+import 'dart:async';
+
+import 'package:http/http.dart' as http;
+
+/// Interceptor that can manipulate http requests and responses.
+abstract interface class HttpInterceptor {
+  /// Whether this interceptor should intercept requests.
+  bool shouldInterceptRequest(http.BaseRequest request);
+
+  /// Intercepts the given [request].
+  ///
+  /// Provided requests are not finalized yet. It is an error for an interceptor
+  /// to request itself.
```

Comment:
> It is an error for an interceptor to request itself.

I'm not sure I understand, you mean the interception callback should not execute the request?